qualtran/bloqs/mcmt/multi_controlled_bloq.py

```diff
@@ -0,0 +1,116 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import TYPE_CHECKING
+
+import numpy as np
+from attrs import frozen
+
+from qualtran import Bloq, Controlled, CtrlSpec
+from qualtran.bloqs.basic_gates import XGate
+from qualtran.bloqs.mcmt.ctrl_spec_and import CtrlSpecAnd
+
+if TYPE_CHECKING:
+    from qualtran import BloqBuilder, Signature, SoquetT
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class MultiControlledBloq(Bloq):
```

Comment:
I see @mpharrigan's point that having both `Controlled(sub_bloq, ctrl_spec)` and `MultiControlledBloq(sub_bloq, ctrl_spec)` can be confusing and the implementation of both would actually be identical. 

Regarding interface vs implementation, it may make things cleaner if we rename `MultiControlledBloq` to be more descriptive of the implementation, like `ControlledViaAndLadder()` and explain that the implementation performs `CtrlSpecAnd` -> `Bloq.controlled()` -> `CtrlSpecAnd.adjoint()` to decompose a bloq controlled by arbitrary `CtrlSpec` to a singly controlled Bloq via an And / And^{dagger} ladder and uses $num_controls$ ancilla qubits. 

Then, this specific implementation can be used by users who want to depend upon this behavior instead of the behavior getting potentially updated if we decide to update the decomposition of `Controlled` Bloq later. 

`MultiControlledPauli` can get removed and we will simply update the call sites to use `ControlledViaAndLadder(ctrl_spec, XGate())`

Reply:
I like the bloq name `ControlledViaAndLadder`, will use that